server/src/main/java/org/elasticsearch/action/bulk/BulkOperation.java

```diff
@@ -168,8 +180,63 @@ private void doRedirectFailures() {
         )) {
             return;
         }
-        Map<ShardId, List<BulkItemRequest>> requestsByShard = drainAndGroupRedirectsByShards(clusterState);
-        executeBulkRequestsByShard(requestsByShard, clusterState, this::completeBulkOperation);
+        Runnable runnable = () -> {
+            // Get new cluster state that includes any potential failure store rollovers.
+            var rolledOverState = observer.setAndGetObservedState();
+            Map<ShardId, List<BulkItemRequest>> requestsByShard = drainAndGroupRedirectsByShards(rolledOverState);
+            executeBulkRequestsByShard(requestsByShard, rolledOverState, this::completeBulkOperation);
+        };
+        rollOverFailureStores(runnable);
+    }
+
+    /**
+     * Send rollover requests for all failure stores that need it. After all requests have completed, we execute the given runnable.
+     * Any failures while rolling over will be added to the {@link BulkItemResponse} entries of the index requests that were redirected to
+     * the failure store that failed to roll over.
+     */
+    private void rollOverFailureStores(Runnable runnable) {
+        // Skip allocation of some objects if we don't need to roll over anything.
+        if (failureStoresToBeRolledOver.isEmpty()) {
+            runnable.run();
+            return;
+        }
+        try (RefCountingRunnable refs = new RefCountingRunnable(runnable)) {
+            for (String failureStore : failureStoresToBeRolledOver) {
+                RolloverRequest rolloverRequest = new RolloverRequest(failureStore, null);
+                rolloverRequest.setIndicesOptions(
+                    IndicesOptions.builder(rolloverRequest.indicesOptions())
+                        .failureStoreOptions(new IndicesOptions.FailureStoreOptions(false, true))
+                        .build()
```

Comment:
My reasoning for doing it like this was that I wanted to keep the default indices options of the `RolloverRequest` and wanted to avoid specifying the default myself (i.e. `IndicesOptions.strictSingleIndexNoExpandForbidClosed()`). All these rollover requests will have the same default indices options, so I'm trying to come up with a way to avoid recreating them, but I'm having trouble coming up with an _elegant_ one. At the same time, I don't expect there to be a lot of failure stores to be rolled over in one request, don't you think?

Reply:
> My reasoning for doing it like this was that I wanted to keep the default indices options of the RolloverRequest and wanted to avoid specifying the default myself (i.e. IndicesOptions.strictSingleIndexNoExpandForbidClosed()). 

What about a local constant, that belongs in this class only?

> All these rollover requests will have the same default indices options, so I'm trying to come up with a way to avoid recreating them, but I'm having trouble coming up with an elegant one. At the same time, I don't expect there to be a lot of failure stores to be rolled over in one request, don't you think?

If the problem is a faulty document probably not, but if the cause of the issue is a problem in the mapping, then it's possible that all the requests will fail :). There are issues that could potentially cause all index operations to fail.

Reply:
Good point, still though a template change could affect a lot of data streams. In any case. You can leave it as is. No strong opinion.